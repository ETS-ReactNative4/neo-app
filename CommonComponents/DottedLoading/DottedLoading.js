import React, { Component } from "react";
import PropTypes from "prop-types";
import { Text } from "react-native";

class DottedLoading extends Component {
  static propTypes = {
    text: PropTypes.string.isRequired,
    numOfDots: PropTypes.number.isRequired,
    textStyle: PropTypes.oneOfType([PropTypes.number, PropTypes.object])
  };

  state = {
    dots: ""
  };
  _interval = {};

  componentDidMount() {
    this._interval = setInterval(() => {
      if (this.state.dots.length < this.props.numOfDots) {
        this.setState({
          dots: this.state.dots + "."
        });
      } else {
        this.setState({ dots: "" });
      }
    }, 100);
  }

  componentWillUnmount() {
    clearInterval(this._interval);
  }

  render() {
    let { textStyle, text } = this.props;
    if (!textStyle) textStyle = {};
    return <Text style={textStyle}>{`${text}${this.state.dots}`}</Text>;
  }
}

export default DottedLoading;
