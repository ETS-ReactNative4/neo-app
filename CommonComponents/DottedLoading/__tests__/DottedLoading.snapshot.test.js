import React from "react";
import renderer from "react-test-renderer";
import DottedLoading from "../DottedLoading";

beforeEach(() => {
  jest.useFakeTimers();
});

afterEach(() => {
  jest.clearAllTimers();
});

test("renders the DottedLoading component correctly", done => {
  const props = {
    text: "just some string",
    numOfDots: 10,
    textStyle: { color: "red" },
    animationSpeed: 200
  };

  const component = <DottedLoading {...props} />;
  let tree;
  tree = renderer.create(component);

  expect(tree.toJSON()).toMatchSnapshot();

  // Travel to 200ms in the future.
  jest.advanceTimersByTime(props.animationSpeed);
  // Re-render the DOM.
  tree.update(component);

  // Test that `setInterval` has been called once.
  expect(setInterval).toHaveBeenCalledTimes(1);
  expect(setInterval).toHaveBeenLastCalledWith(
    expect.any(Function),
    props.animationSpeed
  );

  // Make sure we have 1 dot added after the first `setInterval`.
  expect(tree.toJSON().children[0]).toEqual(props.text + ".");

  // Make sure we have 2 dots added after the second `setInterval`.
  jest.advanceTimersByTime(props.animationSpeed);
  tree.update(component);
  expect(tree.toJSON().children[0]).toEqual(props.text + ".".repeat(2));

  // Make sure we have max number of dots added after the tenth `setInterval`.
  jest.advanceTimersByTime(props.animationSpeed * 8); // 8 here because we've already called advanceTimers twice.
  tree.update(component);
  expect(tree.toJSON().children[0]).toEqual(props.text + ".".repeat(10));

  // Make sure we go back to zero dots added at the eleventh `setInterval`.
  jest.advanceTimersByTime(props.animationSpeed);
  tree.update(component);
  expect(tree.toJSON().children[0]).toEqual(props.text);

  // Unmount and make sure the animation `setInterval` is cleared.
  tree.unmount();
  expect(clearInterval).toHaveBeenCalledTimes(1);

  done();
});
