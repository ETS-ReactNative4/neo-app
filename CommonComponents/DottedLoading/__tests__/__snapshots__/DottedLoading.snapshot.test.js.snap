// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the DottedLoading component correctly 1`] = `
<Text
  style={
    Object {
      "color": "red",
    }
  }
>
  just some string
</Text>
`;
